File: 백준/Gold/14938. 서강그라운드/서강그라운드.py
import heapq
import sys

n, m, r = map(int,input().split())
val = list(map(int, input().split()))
graph = [[] for _ in range(n+1)]
for _ in range(r):
  v1, v2, w = map(int, input().split())
  graph[v1].append((v2,w))
  graph[v2].append((v1,w))

INF = sys.maxsize

def djikstra(start):
  dist = [INF]*(n+1)
  dist[start] = 0
  hq = [(0,start)]

  while hq:
    tot_dist, cv = heapq.heappop(hq)
    if tot_dist > dist[cv]:
      continue

    for nv, n_dist in graph[cv]:
      new_dist = tot_dist + n_dist
      if new_dist < dist[nv]:
        dist[nv] = new_dist
        heapq.heappush(hq, (new_dist, nv))
  return dist

ans = 0

for i in range(1,n+1):
  dist = djikstra(i)
  tot = 0
  for j in range(1, n+1):
    if dist[j] <= m:
      tot += val[j-1]
  ans = max(ans, tot)
print(ans)

File: 백준/Gold/14938. 서강그라운드/서강그라운드.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Scanner;

//TIP To <b>Run</b> code, press <shortcut actionId="Run"/> or
// click the <icon src="AllIcons.Actions.Execute"/> icon in the gutter.
public class Main {
	static class Node {
		int vertex;
		int weight;

		public Node(int vertex, int weight) {
			this.vertex = vertex;
			this.weight = weight;
		}
	}

	static List<Node>[] graph;
	static int[] values;
	static int n;

	static int[] djikstra(int start) {
		int[] dist = new int[n+1];
		Arrays.fill(dist, Integer.MAX_VALUE);
		dist[start] = 0;
		PriorityQueue<Node> pq = new PriorityQueue<>(Comparator.comparingInt(n -> n.weight));
		pq.add(new Node(start, 0));
		while (!pq.isEmpty()) {
			Node node = pq.poll();
			int totDist = node.weight;
			int cv = node.vertex;
			if (dist[cv] < totDist) {
				continue;
			}
			for (Node n: graph[cv]){
				int newDist = totDist + n.weight;
				if (dist[n.vertex] > newDist) {
					dist[n.vertex] = newDist;
					pq.add(new Node(n.vertex, newDist));
				}
			}
		}
		return dist;
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		n = sc.nextInt();
		int m = sc.nextInt();
		int r = sc.nextInt();

		values = new int[n];
		for (int i = 0; i < n; i++) {
			values[i] = sc.nextInt();
		}

		graph = new ArrayList[n + 1];
		for (int i = 0; i < n + 1; i++) {
			graph[i] = new ArrayList<>();
		}
		for (int i = 0; i < r; i++) {
			int v1 = sc.nextInt();
			int v2 = sc.nextInt();
			int w = sc.nextInt();
			graph[v1].add(new Node(v2, w));
			graph[v2].add(new Node(v1, w));
		}

		int ans = 0;
		for(int i = 1; i <= n; i++){
			int[] dist = djikstra(i);
			int tot = 0;
			for(int j = 1; j <= n; j++){
				if (dist[j] < m + 1) {
					tot += values[j-1];
				}
			}
			ans = Math.max(ans, tot);
		}
		System.out.println(ans);
	}
}